### PR TITLE
BenchMark library Updated

### PR DESCRIPTION
**Pull request by mehul11jain**

### CODE REVIEW

**Comment by rksharma-yymd** [comment]
Many thanks! 

@mehul11jain Can we have a benchmark interface and for every benchmark we create a derived class with benchmark-specific details? The present benchmark logic is targeting PTABen, it can be kept in a separate file and so can be the logic for your benchmark.

**Comment by reSHARMA** [comment]
Many thanks! 

@mehul11jain Can we have a benchmark interface and for every benchmark, we create a derived class with benchmark-specific details? The present benchmark logic is targeting PTABen, it can be kept in a separate file and so can be the logic for your benchmark.

**Comment by mehul11jain** [comment]
Sure we will create a separate child class "PTABenchmarkRunner" of the BenchmarkRunner class. So that your implementation for PTABen test cases would still work and our implementation is separate from the original implementation.

**Comment by mehul11jain** [comment]
Made a separate class PTABenchmarkRunner which is a derived class of the BenchmarkRunner class. Please have a look.

**Comment by reSHARMA** [comment]
Thanks but I was thinking something like having benchmark runner class as an interface. PTABen and your benchmark can be derived from this interface. You also need to name your benchmark something else (PTABen already exists). 
This will require changes on the "client" analysis side but I think that is fine. It gives more flexibility to the client analysis to pick the benchmark runner they want to use.

**Comment by mehul11jain** [comment]
Currently, there are two classes in the Benchmark library.
* **BenchmarkRunner** (which was already implemented).
* **PTABenchmarkRunner** class that we created to evaluate the results of the analysis on the test cases created by us.

So should we create an abstract class let's say "BenchmarkRunnerInterface" with the extract function being pure virtual and Both these (**BenchmarkRunner** and **PTABenchmarkRunner**) classes inheriting the interface and implementing the extract function and adding some other functions specific to them? Is this what is needed?

**Comment by reSHARMA** [comment]
You are right but your test cases/benchmarks use predicates that are not present in PTABen so they should be considered as a separate benchmark (and you should call it something). We can have BenchmarkRunner as the interface and PTABenchmarkRunner, YourBenchmarkRunner and etc can be derived from it. 
I'm trying not to make PTABen the default benchmark that we support.

**Comment by mehul11jain** [comment]
Created an abstract class **BenchmarkRunner** which is derived by two subclasses.
* PTABenchmarkRunner: Contains the original code which was written in Benchmark.cpp by you;
* PointsToBenchmarkRunner: Contains the code written by us to evaluate our test cases. 

Is this what was needed or am I missing something? Kindly check.

File: include/Benchmark/Benchmark.h [hunk 1]
```diff
@@ -12,30 +12,14 @@
...
   std::string Key;
-  std::map<llvm::Instruction *, std::string> Result;
-  void evaluateUtil(llvm::CallInst *, int, int, int);
 
 public:
-  BenchmarkRunner() { Key = "ALIAS"; }
-  std::vector<llvm::Value *> extract(llvm::Instruction *Inst);
-  friend std::ostream &operator<<(std::ostream &OS, const BenchmarkRunner &B);
-  template <class Ty>
-  void evaluate(llvm::Instruction *Inst, std::set<Ty *>, std::set<Ty *>);
```
Comment (reSHARMA) [inline include/Benchmark/Benchmark.h:25]: Will the interface be incomplete without evaluate?

Comment (mehul11jain) [reply include/Benchmark/Benchmark.h:25]: Evaluate function has different signatures for both PTABenchmarkRunner and PointsToBenchmarkRunner that why we implemented evaluate function separately for both the classes.

File: include/Benchmark/PointsToBenchmark.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  std::vector<llvm::Value *> extract(llvm::Instruction *Inst);
+  friend std::ostream &operator<<(std::ostream &OS, const PointsToBenchmarkRunner &B);
+  template <class Ty>
+  void evaluate(llvm::Instruction *Inst, std::set<Ty *>, Ty *);
+  spatial::Soundness CheckSoundness();
+  void evaluatePrecision(llvm::Function &);
+};
+
+template <class Ty>
+void PointsToBenchmarkRunner::evaluate(llvm::Instruction *I, std::set<Ty *> A,
```
Comment (reSHARMA) [inline include/Benchmark/PointsToBenchmark.h:31]: Can you add some comments here to let the analysis writer know about your expectation wrt the arguments of evaluate? 
You may want to add Doxygen-friendly comments (https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments).

File: include/Benchmark/PointsToBenchmark.h [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+                                     Ty *B) {
+  llvm::CallInst *Inst = llvm::cast<llvm::CallInst>(I);
+  std::string Status, Expected;
+  llvm::StringRef FunctionName = Inst->getCalledFunction()->getName();
+
+  llvm::MDNode *md = I->getMetadata("dbg");
```
Comment (reSHARMA) [inline include/Benchmark/PointsToBenchmark.h:37]: nit, md ->MD

File: include/Benchmark/PointsToBenchmark.h [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  int lineno = llvm::cast<llvm::DILocation>(md)->getLine();
```
Comment (reSHARMA) [inline include/Benchmark/PointsToBenchmark.h:38]: nit, lineno -> LineNo

File: include/Benchmark/PointsToBenchmark.h [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+  // std::cout << lineno << "\n";
```
Comment (reSHARMA) [inline include/Benchmark/PointsToBenchmark.h:40]: nit, you may want to remove this

File: lib/Benchmark/PointsToBenchmark.cpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+#include <iomanip>
+#define CYANB "\033[1;46m"
+#define REDT "\033[1;91m"
+#define GREENT "\033[1;92m"
+#define YELLOWT "\033[1;93m"
+#define RST "\033[0;m"
+
+namespace spatial {
+
+std::string print_args(llvm::CallInst* ci){
```
Comment (reSHARMA) [inline lib/Benchmark/PointsToBenchmark.cpp:21]: nit, ci -> CI

File: lib/Benchmark/PointsToBenchmark.cpp [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    int i = 1;
+    std::string header = "*********************************************** BENCHMARK RESULTS *********************************************************",res;
+    OS << "\n__________________________________________________________________________________________________________________________\n";
+    OS << header + "\n";
+    OS << "---------------------------------------------------------------------------------------------------------------------------\n\n";    
+    for (auto R = B.TCResult.begin();R != B.TCResult.end();R++) {
+        if(R == B.TCResult.begin()){
+            OS << "\t\t=================================================================================\n";
+        }
+        llvm::CallInst* ci = llvm::dyn_cast<llvm::CallInst>((*R).first);
```
Comment (reSHARMA) [inline lib/Benchmark/PointsToBenchmark.cpp:57]: nit, ci -> CI

File: include/Benchmark/PointsToBenchmark.h [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+#include "llvm/IR/DebugInfoMetadata.h"
+
+namespace spatial {
+
+enum Soundness {
+  Sound,
+  Unsound,
+};
+class PointsToBenchmarkRunner : public BenchmarkRunner {
+  Soundness SoundOrUnsound;
```
Comment (reSHARMA) [inline include/Benchmark/PointsToBenchmark.h:14]: SoundOrUnsound, can we rename this? Is it required to have it this way? How about, bool IsSound?

Comment (mehul11jain) [reply include/Benchmark/PointsToBenchmark.h:14]: Sure, we can change this. We just created an enum for readability.

**Review by reSHARMA — CHANGES_REQUESTED** [review]
The test cases need to go into a separate repo :) (like PTABen). I've added a few comments. Everything else looks good to me. I recall that PTABen evaluate function was buggy? Do you have a fixed version for that, if not, can you please create an issue.

**Comment by mehul11jain** [comment]
Made the changes.

**Comment by reSHARMA** [comment]
Many thanks! Looks good to me!

**Pull request merged**